### ISSUE TITLE
Question: Best way to deal with returning a big message?

### ISSUE THREAD

**Issue by amamh** [dotnet/orleans#1481]
You can skip to the end to save time.

It seems like the server takes a very long time when it has to send a large message.

Here is an example:

      [2016-02-17 17:51:37.369 GMT    24	INFO   	101018	Message	169.183.186.58:0]	Preparing to send large message Size=288414 HeaderLength=251 BodyLength=288155 #ArraySegments=74. Msg=Request *cli/0984e7a1@50b9f88b->S169.183.186.58:30000:0*grn/B9FC61B7/00000000 #5: global::Server.GrainInterface.IInstrumentQueryGrain:Add()	

Size here is 288155 which is less than 300 KB, so really not that big. I'm also running everything on the same PC. However, I remember it actually taking more than 5 seconds to send which seems extremely long to me considering the size.

I solved this problem when sending to a grain by **sending the message in batches**. My message was a list of items, so it was easy to just split them into smaller lists.

The problem now is that I don't see how this can be done when it's the other way around i.e. **a grain is returning a big message to its caller**.

--------

**TL;DR**
1- Am I supposed to send only small messages?
2- Why does it take so long to send a 300 KB message on the same PC?
3- How to handle this when returning a big message from a grain to the caller?

**Comment by sergeybykov** [comment]
This is very strange. There is no restriction on sending small messages, and of course 300KB cannot take even close to that long. The warning is for information purposes and is a bit historical. It was more important before we implemented the buffer pool and Large Object Heap was an issue.

Can you double-check what you are measuring? Is it possible that you included the grain activation time, and the grain is somehow slow in its OnActivateAsync()? Just guessing. Or maybe there is some other one-time initialization that's happening upon a first call.

**Comment by amamh** [comment]
@sergeybykov 

Thanks for the reply, I'll do further testing. If I can get a sample project to reproduce the problem I'll post it here.

**Comment by sergeybykov** [comment]
:+1:
To be clear, many of Orleans-based services send large messages. Even internally, cache validator sends huge messages to directory partitions. And it's always fast.

**Comment by amamh** [comment]
I think this could have been a mistake on my part somehow, but unfortunately I cannot investigate further at the moment.
